File: 21-merge-two-sorted-lists/merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        
        newHead = ListNode(-1)
        tmp = newHead

        l1 = list1
        l2 = list2

        while ( l1 and l2):
            if ( l1.val < l2.val):
                newNode = ListNode(l1.val)
                tmp.next = newNode
                tmp=tmp.next
                l1 = l1.next
            else:
                newNode = ListNode(l2.val)
                tmp.next = newNode
                tmp=tmp.next
                l2 = l2.next

        while l2:
            remNode = ListNode(l2.val)
            tmp.next = remNode
            tmp = tmp.next
            l2 = l2.next
        
        while l1:
            remNode = ListNode(l1.val)
            tmp.next = remNode
            tmp = tmp.next
            l1 = l1.next
        
        return newHead.next

        

File: 21-merge-two-sorted-lists/merge-two-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) 
    {
        if (!list1) return list2;
        if (!list2) return list1;

        ListNode* dummy = new ListNode(-1);
        ListNode* l3 = dummy;

        ListNode* l1 = list1;
        ListNode* l2 = list2;

        while ( l1 && l2)
        {
            if ( l1->val < l2->val)
            {
                ListNode* valNode = new ListNode(l1->val);
                l3->next = valNode;
                l3 = l3->next;
                l1 = l1->next;
            }
            else
            {
                ListNode* valNode = new ListNode(l2->val);
                l3->next = valNode;
                l3 = l3->next;
                l2 = l2->next;

            }
        }

        while ( l1)
        {
            l3->next = l1;
            l1 = l1->next;
            l3 = l3->next;
        }

        while ( l2)
        {
            l3->next = l2;
            l2 = l2->next;
            l3 = l3->next;
        }

        return dummy->next;



        
    }
};